Android/backend/intra/internal/doh/stream_dialer.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package doh
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"net/netip"
+	"sync/atomic"
+	"time"
+
+	"github.com/Jigsaw-Code/Intra/Android/backend/intra/internal/sni"
+	intraLegacy "github.com/Jigsaw-Code/outline-go-tun2socks/intra"
+	"github.com/Jigsaw-Code/outline-go-tun2socks/intra/doh"
+	"github.com/Jigsaw-Code/outline-go-tun2socks/intra/protect"
+	"github.com/Jigsaw-Code/outline-go-tun2socks/intra/split"
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+)
+
+type DoHStreamDialer interface {
+	transport.StreamDialer
+
+	SetDoHTransport(DoHTransport) error
+}
+
+type tcpTrafficStats = intraLegacy.TCPSocketSummary
+
+type dohSplitStreamDialer struct {
+	fakeDNSAddr      netip.AddrPort
+	dohServer        atomic.Pointer[DoHTransport]
+	dialer           *net.Dialer
+	alwaysSplitHTTPS atomic.Bool
+	listener         intraLegacy.TCPListener
+	sniReporter      sni.TCPSNIReporter
+}
+
+var _ DoHStreamDialer = (*dohSplitStreamDialer)(nil)
+
+func MakeDoHStreamDialer(fakeDNS netip.AddrPort, dohServer DoHTransport, protector Protector, listener intraLegacy.TCPListener, sniReporter sni.TCPSNIReporter) (DoHStreamDialer, error) {
```

Comment:
This isn't a DoH StreamDialer. It's an `IntraStreamDialer`. It does a lot more than DoH. We need to rename it.